### PR TITLE
Max texture size calculator

### PR DESCRIPTION
**Pull request by Legonaftik**

### CODE REVIEW

File: Alloy/MTLDevice+Extensions.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        return false
+        #endif
+    }
+}
+
+/// MTLTexture cannot be created if the size is very close to the documented limit.
+/// We make images smaller to make sure they can be processed.
+/// It also avoids memory issues.
+private let metalSafeMultiplier: CGFloat = 0.75
```
Comment (s1ddok) [inline Alloy/MTLDevice+Extensions.swift]: I think it should be moved to a local scope

File: Alloy/MTLDevice+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import Metal
+import CoreGraphics
+
+@available(iOS 11.0, macOS 10.12, *)
+public extension MTLDevice {
+
+    func maxTextureSize(desiredSize: CGSize) -> CGSize {
+        let maxSide: CGFloat
+        if self.supportsOnly8K() {
+            maxSide = 8192 * metalSafeMultiplier
```
Comment (s1ddok) [inline Alloy/MTLDevice+Extensions.swift]: I think safeness should be define by the app, not the core framework

File: Alloy/MTLDevice+Extensions.swift [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+//  Created by Vladimir Pavlov on 06/05/2019.
+//
+
+import Metal
+import CoreGraphics
+
+@available(iOS 11.0, macOS 10.12, *)
+public extension MTLDevice {
+
+    func maxTextureSize(desiredSize: CGSize) -> CGSize {
```
Comment (s1ddok) [inline Alloy/MTLDevice+Extensions.swift]: I think it is better to use MTLSize

File: Alloy/MTLDevice+Extensions.swift [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        } else {
+            let resultHeight = min(desiredSize.height, maxSide)
+            let resultWidth = resultHeight * aspectRatio
+            return CGSize(width: resultWidth, height: resultHeight)
+        }
+    }
+
+    private func supportsOnly8K() -> Bool {
+        #if os(macOS)
+        return false
```
Comment (s1ddok) [inline Alloy/MTLDevice+Extensions.swift]: that's temporary of an actual fact?

File: Alloy/MTLDevice+Extensions.swift [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        #else
+        if self.supportsFeatureSet(.iOS_GPUFamily1_v4) { return true }
+        if self.supportsFeatureSet(.iOS_GPUFamily2_v4) { return true }
+        if #available(iOS 12.0, *) {
+            if self.supportsFeatureSet(.iOS_GPUFamily1_v5) { return true }
```
Comment (s1ddok) [inline Alloy/MTLDevice+Extensions.swift]: i think that's redundant

**Comment by s1ddok** [comment]
@eugenebokhan please review this

File: Alloy/MTLDevice+Extensions.swift [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    func maxTextureSize(desiredSize: CGSize) -> CGSize {
+        let maxSide: CGFloat
+        if self.supportsOnly8K() {
+            maxSide = 8192 * metalSafeMultiplier
+        } else {
+            maxSide = 16_384 * metalSafeMultiplier
+        }
+
+        guard desiredSize.width > 0, desiredSize.height > 0 else { return .zero }
```
Comment (eugenebokhan) [inline Alloy/MTLDevice+Extensions.swift]: I would change style a little bit:
```swift
guard 
    desiredSize.width > 0, 
    desiredSize.height > 0 
else { return .zero }
```
But it's optional, up to you

File: Alloy/MTLDevice+Extensions.swift (hunk 4 above)
Comment (Legonaftik) [reply Alloy/MTLDevice+Extensions.swift]: All Mac versions that Alloy supports can create 16K textures.

File: Alloy/MTLDevice+Extensions.swift (hunk 5 above)
Comment (Legonaftik) [reply Alloy/MTLDevice+Extensions.swift]: Why? E.g., iPhone 5S and iPhone 6 support iOS 12.

Comment (s1ddok) [reply Alloy/MTLDevice+Extensions.swift]: yup, but that's cumulative. if the device supporst GPUFamily1_v5 it automatically supports GPUFamily1_v[N] N < 5

we discussed that in person I think

**Comment by s1ddok** [comment]
@Legonaftik still would love to see only .swift files in this PR

**Pull request merged**